collector/taskrun.go

```diff
@@ -0,0 +1,83 @@
+package collector
+
+import (
+	"context"
+	"fmt"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type ReconcileTaskRun struct {
+	client        client.Client
+	scheme        *runtime.Scheme
+	eventRecorder record.EventRecorder
+	trCollector   *TaskRunCollector
+}
+
+type trStartTimeEventFilter struct {
+}
+
+func (f *trStartTimeEventFilter) Create(event.CreateEvent) bool {
+	return false
+}
+
+func (f *trStartTimeEventFilter) Generic(event.GenericEvent) bool {
+	return false
+}
+
+func (f *trStartTimeEventFilter) Delete(event.DeleteEvent) bool {
+	return false
+}
+
+func (f *trStartTimeEventFilter) Update(e event.UpdateEvent) bool {
+	//TODO remember, keep track of when pipeline-service and RHTAP starts moving from v1beta1 to v1
+	oldTR, okold := e.ObjectOld.(*v1beta1.TaskRun)
+	newTR, oknew := e.ObjectNew.(*v1beta1.TaskRun)
+	if okold && oknew {
+		if oldTR.Status.StartTime == nil && newTR.Status.StartTime != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func SetupTaskRunController(mgr ctrl.Manager) error {
+	reconciler := &ReconcileTaskRun{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		eventRecorder: mgr.GetEventRecorderFor("MetricExporterTaskRuns"),
+		trCollector:   NewTaskRunCollector(),
+	}
+	return ctrl.NewControllerManagedBy(mgr).For(&v1beta1.TaskRun{}).WithEventFilter(&trStartTimeEventFilter{}).Complete(reconciler)
+}
+
+func (r *ReconcileTaskRun) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```

Comment:
Do you think adding something like (with maybe a higher verbosity level)-
```
log.V(7).Info("Starting reconciliation", "namespace", request.Namespace, "name", request.Name)
```
will make sense or will it still be a lot of noise?

Reply:
yeah I don't think a log like this for a basic "did we get events" notification really helps

by the way, there is already a controller runtime stat that provides this:  see `controller_runtime_reconcile_total` over at https://book.kubebuilder.io/reference/metrics-reference.html

Reply:
Sure, that makes sense.